### PR TITLE
reduces memory footprint of schema

### PR DESCRIPTION
**Pull request by PascalSenn**
Reduces memory footprint of schema

### CODE REVIEW

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Definitions/IHasDirectiveDefinition.cs [hunk 1]
```diff
@@ -9,5 +9,11 @@ public interface IHasDirectiveDefinition
         /// the implementing object.
         /// </summary>
         IList<DirectiveDefinition> Directives { get; }
+
+        /// <summary>
+        /// Gets the list of directives that are annotated to
+        /// the implementing object.
+        /// </summary>
+        IReadOnlyList<DirectiveDefinition> GetDirectives();
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types/Types/Descriptors/Definitions/IHasDirectiveDefinition.cs:17]: This is a breaking change on the interface!

File: src/HotChocolate/Core/src/Types/Types/InputField.cs [hunk 2]
```diff
@@ -23,11 +23,11 @@ public InputField(InputFieldDefinition definition, FieldCoordinate fieldCoordina
             DefaultValue = definition.DefaultValue;
             Property = definition.Property;
 
-            if (definition.Formatters.Count == 0)
+            if (definition.GetFormatters().Count == 0)
             {
                 Formatter = null;
             }
-            else if (definition.Formatters.Count == 1)
+            else if (definition.GetFormatters().Count == 1)
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/InputField.cs:30]: var formatters = definition.GetFormatters();

if(formatters.Count == 0)
and so on....

File: src/HotChocolate/Core/src/Types/Types/InterfaceCompletionTypeInterceptor.cs [hunk 3]
```diff
@@ -11,12 +11,12 @@ namespace HotChocolate.Types
 {
     internal class InterfaceCompletionTypeInterceptor : TypeInterceptor
     {
-        private readonly Dictionary<ITypeSystemObject, TypeInfo> _typeInfos = new();
-        private readonly HashSet<Type> _allInterfaceRuntimeTypes = new();
-        private readonly HashSet<Type> _interfaceRuntimeTypes = new();
-        private readonly HashSet<NameString> _completed = new();
-        private readonly HashSet<NameString> _completedFields = new();
-        private readonly Queue<InterfaceType> _backlog = new();
+        private Dictionary<ITypeSystemObject, TypeInfo> _typeInfos = new();
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/InterfaceCompletionTypeInterceptor.cs:14]: why not readonly?

File: src/HotChocolate/Core/src/Types/Types/InterfaceCompletionTypeInterceptor.cs [hunk 4]
```diff
@@ -149,9 +149,20 @@ private Type GetRuntimeType(TypeInfo typeInfo)
             }
         }
 
+        public override void OnTypesCompleted(
+            IReadOnlyCollection<ITypeCompletionContext> completionContexts)
+        {
+            _typeInfos = null!;
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/InterfaceCompletionTypeInterceptor.cs:155]: I mean the whole type interceptor is thrown away afterwards...

File: src/HotChocolate/Core/src/Types/Types/ObjectField.cs [hunk 5]
```diff
@@ -19,7 +19,7 @@ public class ObjectField
         , IObjectField
     {
         private static readonly FieldDelegate _empty = _ => throw new InvalidOperationException();
-        private readonly List<IDirective> _executableDirectives = new();
+        private IReadOnlyList<IDirective> _executableDirectives = Array.Empty<IDirective>();
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/ObjectField.cs:22]: Why readonly list? why not an array?

File: src/HotChocolate/Core/src/Types/Types/InterfaceType.cs [hunk 6]
```diff
@@ -17,7 +17,7 @@ public class InterfaceType
         : NamedTypeBase<InterfaceTypeDefinition>
         , IInterfaceType
     {
-        private readonly List<InterfaceType> _implements = new();
+        private IReadOnlyList<InterfaceType> _implements = Array.Empty<InterfaceType>();
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/InterfaceType.cs:20]: Why readonly list? why not an array?

File: src/HotChocolate/Core/src/Types/Types/ObjectType.cs [hunk 7]
```diff
@@ -19,7 +19,7 @@ public class ObjectType
         : NamedTypeBase<ObjectTypeDefinition>
         , IObjectType
     {
-        private readonly List<InterfaceType> _implements = new();
+        private IReadOnlyList<InterfaceType> _implements = Array.Empty<InterfaceType>();
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/ObjectType.cs:22]: Why readonly list? why not an array?

**Comment by michaelstaib** [comment]
Why are so many tests breaking?

**Comment by PascalSenn** [comment]
@michaelstaib  I do not really know why the strawberry shake tests are failing. Could these be related to the syntax token that is no longer on the field?

File: src/HotChocolate/Core/src/Types/SchemaBuilder.Lazy.cs [hunk 8]
```diff
@@ -36,6 +36,7 @@ public ISchema Schema
                     _schema = value ?? throw new ArgumentNullException(nameof(value));
                     _isSet = true;
                     Completed?.Invoke(this, EventArgs.Empty);
+                    Completed = null;
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types/SchemaBuilder.Lazy.cs:39]: @michaelstaib  this is where the definitions were kept alive

File: src/HotChocolate/Core/src/Types/Types/FieldCollection.cs [hunk 9]
```diff
@@ -24,10 +24,12 @@ public FieldCollection(IEnumerable<T> fields, bool sortByName = false)
             }
 
             _fields = sortByName
-                ? fields.OrderBy(t => t.Name).ToList()
-                : fields is List<T> list ? list : fields.ToList();
+                ? fields.OrderBy(t => t.Name).ToArray()
+                : fields is T[] list ? list : fields.ToArray();
+
+            _fieldsLookup = new Dictionary<NameString, (int Index, T Field)>(_fields.Length);
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types/Types/FieldCollection.cs:30]: Is this a good idea?

File: src/HotChocolate/Core/src/Types/InternalsVisibleTo.cs [hunk 10]
```diff
@@ -14,3 +14,4 @@
 [assembly: InternalsVisibleTo("HotChocolate.Data.Filters.Tests")]
 [assembly: InternalsVisibleTo("HotChocolate.Data.Sorting.Tests")]
 [assembly: InternalsVisibleTo("HotChocolate.Data.Projections.Tests")]
+[assembly: InternalsVisibleTo("HotChocolate.Stitching")]
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/InternalsVisibleTo.cs:17]: No more internal visibles.

File: src/HotChocolate/Core/src/Types/SchemaBuilder.Lazy.cs (hunk 8 above)
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types/SchemaBuilder.Lazy.cs:39]: good catch

File: src/HotChocolate/Core/benchmark/Execution.Benchmarks/reports/HotChocolate.Execution.Benchmarks.DefaultExecutionPipelineBenchmark-report-github.md [hunk 11]
```diff
@@ -1,22 +1,22 @@
...
+|                GetHeroFiveParallelRequests |  2.545 ms | 0.0516 ms | 0.1522 ms |  2.525 ms |    3 |    3.9063 |        - |        - |    41.04 KB |
+|                         GetHeroWithFriends |  7.600 ms | 0.0831 ms | 0.0924 ms |  7.593 ms |    5 |         - |        - |        - |    82.35 KB |
+|     GetHeroWithFriendsFiveParallelRequests |  8.239 ms | 0.0953 ms | 0.0796 ms |  8.264 ms |    6 |   31.2500 |  15.6250 |        - |    438.4 KB |
+|                     GetTwoHerosWithFriends |  8.192 ms | 0.1781 ms | 0.5251 ms |  7.965 ms |    6 |   15.6250 |        - |        - |   178.88 KB |
+| GetTwoHerosWithFriendsFiveParallelRequests |  9.350 ms | 0.1801 ms | 0.4139 ms |  9.224 ms |    7 |   78.1250 |  31.2500 |        - |   911.82 KB |
+|                                 LargeQuery | 17.720 ms | 0.3495 ms | 0.6303 ms | 17.664 ms |    8 |  312.5000 | 156.2500 |  31.2500 |  3299.39 KB |
+|             LargeQueryFiveParallelRequests | 40.000 ms | 0.7795 ms | 0.9573 ms | 39.892 ms |    9 | 1538.4615 | 692.3077 | 153.8462 | 16156.59 KB |
```
Comment (PascalSenn) [inline src/HotChocolate/Core/benchmark/Execution.Benchmarks/reports/HotChocolate.Execution.Benchmarks.DefaultExecutionPipelineBenchmark-report-github.md:22]: Gen 2?

File: src/HotChocolate/Core/src/Types/Types/FieldCollection.cs (hunk 9 above)
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types/Types/FieldCollection.cs:30]: This is the initial capacity 
the dict can always expand.

File: src/HotChocolate/Core/src/Types/Types/InterfaceType.cs [hunk 12]
```diff
@@ -131,7 +132,21 @@ protected virtual void Configure(IInterfaceTypeDescriptor descriptor)
                 sortFieldsByName);
 
             CompleteAbstractTypeResolver(context, definition.ResolveAbstractType);
-            CompleteInterfaces(context, definition, RuntimeType, _implements, this, SyntaxNode);
+
+            if (definition.GetInterfaces().Any())
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/InterfaceType.cs:136]: tzzz

**Pull request merged**